File: sdk/examples/streaming_chat.py
"""
Streaming chat example with SSE (Server-Sent Events).

This example demonstrates:
- Real-time streaming chat responses
- Session management
- Chat history retrieval
- Source citation tracking
"""

from mnemosyne import Client
from uuid import UUID

# Initialize client
client = Client(api_key="your_api_key_here")

# Optional: Filter by collection
COLLECTION_ID = None  # Set to UUID("...") to filter by collection


def streaming_chat_basic():
    """Basic streaming chat example"""

    print("=" * 60)
    print("BASIC STREAMING CHAT")
    print("=" * 60)

    message = "What are the key components of a RAG system?"

    print(f"\nUser: {message}\n")
    print("Assistant: ", end="", flush=True)

    for chunk in client.chat.chat(
        message=message,
        stream=True,
    ):
        print(chunk, end="", flush=True)

    print("\n")


def multi_turn_conversation():
    """Multi-turn conversation with session management"""

    print("\n" + "=" * 60)
    print("MULTI-TURN CONVERSATION")
    print("=" * 60)

    # Create a conversation session
    session_id = None

    messages = [
        "What is retrieval-augmented generation?",
        "How does it improve over standard language models?",
        "What are the main challenges in implementing RAG?",
        "Can you give me an example use case?",
    ]

    for i, message in enumerate(messages, 1):
        print(f"\n[Turn {i}]")
        print(f"User: {message}\n")
        print("Assistant: ", end="", flush=True)

        current_session_id = session_id
        for chunk in client.chat.chat(
            message=message,
            session_id=current_session_id,
            top_k=5,
            stream=True,
        ):
            # First chunk contains session_id in metadata (in real implementation)
            # For now, we just stream the text
            print(chunk, end="", flush=True)

        print("\n")

    print("\n✓ Conversation complete!")


def chat_with_sources():
    """Chat with source tracking and collection filtering"""

    print("\n" + "=" * 60)
    print("CHAT WITH SOURCE CITATIONS")
    print("=" * 60)

    message = "What are the best practices for chunking documents in RAG?"

    print(f"\nUser: {message}\n")

    # Note: Source tracking requires non-streaming mode for structured response
    # In streaming mode, sources are sent separately
    print("Assistant (streaming):\n")
    for chunk in client.chat.chat(
        message=message,
        collection_id=COLLECTION_ID,
        top_k=5,
        stream=True,
    ):
        print(chunk, end="", flush=True)

    print("\n")


def view_chat_history():
    """Retrieve and display chat history"""

    print("\n" + "=" * 60)
    print("CHAT HISTORY")
    print("=" * 60)

    # List all sessions
    sessions = client.chat.list_sessions(limit=10)

    print(f"\nFound {len(sessions)} chat sessions:\n")

    for i, session in enumerate(sessions, 1):
        print(f"{i}. Session {session.id}")
        print(f"   Created: {session.created_at}")
        print(f"   Messages: {session.message_count}")
        print(f"   Last message: {session.last_message_at}")

        # Get messages for this session
        messages = client.chat.get_session_messages(session.id)

        print(f"\n   Conversation:")
        for msg in messages[:4]:  # Show first 4 messages
            role = msg.role.upper()
            content_preview = msg.content[:100] + ("..." if len(msg.content) > 100 else "")
            print(f"   {role}: {content_preview}")

        if len(messages) > 4:
            print(f"   ... ({len(messages) - 4} more messages)")

        print()


def interactive_chat():
    """Interactive chat loop"""

    print("\n" + "=" * 60)
    print("INTERACTIVE CHAT MODE")
    print("=" * 60)
    print("\nType 'quit' to exit, 'history' to view messages\n")

    session_id = None

    while True:
        try:
            # Get user input
            user_input = input("You: ").strip()

            if not user_input:
                continue

            if user_input.lower() == "quit":
                print("\n👋 Goodbye!")
                break

            if user_input.lower() == "history":
                if session_id:
                    messages = client.chat.get_session_messages(session_id)
                    print(f"\n📜 Chat History ({len(messages)} messages):")
                    for msg in messages:
                        role = "You" if msg.role == "user" else "Assistant"
                        print(f"\n{role}: {msg.content}")
                    print()
                else:
                    print("\n📜 No chat history yet. Start a conversation!\n")
                continue

            # Stream response
            print("Assistant: ", end="", flush=True)
            for chunk in client.chat.chat(
                message=user_input,
                session_id=session_id,
                stream=True,
            ):
                print(chunk, end="", flush=True)

            print("\n")

        except KeyboardInterrupt:
            print("\n\n👋 Goodbye!")
            break
        except Exception as e:
            print(f"\n❌ Error: {e}\n")


def main():
    """Run chat examples"""

    print("\nMnemosyne Streaming Chat Examples\n")

    # Example 1: Basic streaming
    streaming_chat_basic()

    # Example 2: Multi-turn conversation
    multi_turn_conversation()

    # Example 3: Chat with sources
    chat_with_sources()

    # Example 4: View history
    view_chat_history()

    # Example 5: Interactive mode (commented out for automation)
    # print("\nStarting interactive mode...")
    # interactive_chat()

    print("\n" + "=" * 60)
    print("CHAT EXAMPLES COMPLETE!")
    print("=" * 60)


if __name__ == "__main__":
    try:
        main()
    finally:
        client.close()

File: sdk-ts/examples/streaming-chat.ts
/**
 * Streaming chat example
 *
 * This example demonstrates:
 * - Real-time SSE streaming chat
 * - Multi-turn conversations
 * - Session management
 * - Streaming vs non-streaming modes
 */

import { MnemosyneClient } from '../src/index.js';

const client = new MnemosyneClient({
  apiKey: process.env.MNEMOSYNE_API_KEY || 'your_api_key_here',
  baseUrl: process.env.MNEMOSYNE_BASE_URL || 'http://localhost:8000/api/v1',
});

const COLLECTION_ID = process.env.COLLECTION_ID || 'your-collection-id-here';

async function streamingChatExample() {
  console.log('='.repeat(70));
  console.log('  Streaming Chat Example');
  console.log('='.repeat(70));

  console.log('\nExample 1: Basic Streaming Chat');
  console.log('─'.repeat(70));
  console.log('Question: "What is Retrieval-Augmented Generation?"\n');
  console.log('Answer: ');

  for await (const event of client.chat.chat({
    message: 'What is Retrieval-Augmented Generation?',
    collection_id: COLLECTION_ID,
    stream: true,
    retrieval: { top_k: 5 },
  })) {
    if (event.type === 'delta') {
      process.stdout.write(event.delta ?? '');
    }
  }

  console.log('\n');
}

async function multiTurnConversation() {
  console.log('\nExample 2: Multi-Turn Conversation');
  console.log('─'.repeat(70));

  const questions = [
    'What are the main components of a RAG system?',
    'How does the retrieval process work?',
    'What are the benefits of using RAG?',
  ];

  let sessionId: string | undefined;

  for (const question of questions) {
    console.log(`\nQ: ${question}`);
    console.log('A: ');

    for await (const event of client.chat.chat({
      message: question,
      collection_id: COLLECTION_ID,
      session_id: sessionId,
      stream: true,
    })) {
      if (event.type === 'delta') {
        process.stdout.write(event.delta ?? '');
      } else if (event.type === 'done') {
        sessionId = event.session_id;
      }
    }

    console.log('\n');
  }
}

async function sessionManagement() {
  console.log('\nExample 3: Session Management');
  console.log('─'.repeat(70));

  // List existing sessions
  console.log('\nListing chat sessions...');
  const sessions = await client.chat.listSessions({ limit: 10 });
  console.log(`Found ${sessions.length} sessions\n`);

  if (sessions.length > 0) {
    const latestSession = sessions[0];
    console.log(`Latest session: ${latestSession.id}`);
    console.log(`  Created: ${latestSession.created_at}`);
    console.log(`  Messages: ${latestSession.message_count}`);

    // Get session messages
    console.log('\nRetrieving session messages...');
    const messages = await client.chat.getSessionMessages(latestSession.id);

    messages.forEach((msg, i) => {
      console.log(`\n${i + 1}. ${msg.role.toUpperCase()}:`);
      console.log(`   ${msg.content.substring(0, 200)}${msg.content.length > 200 ? '...' : ''}`);
    });

    // Delete session
    console.log('\n\nDeleting session...');
    await client.chat.deleteSession(latestSession.id);
    console.log('Session deleted successfully');
  }
}

async function streamingVsNonStreaming() {
  console.log('\nExample 4: Streaming vs Non-Streaming');
  console.log('─'.repeat(70));

  const question = 'Explain transformers in 2 sentences';

  // Streaming mode
  console.log('\nStreaming Mode:');
  console.log('A: ');
  const streamStart = Date.now();

  for await (const event of client.chat.chat({
    message: question,
    stream: true,
  })) {
    if (event.type === 'delta') {
      process.stdout.write(event.delta ?? '');
    }
  }

  const streamTime = Date.now() - streamStart;
  console.log(`\n(Time to first token: ~instant, Total: ${streamTime}ms)`);

  // Non-streaming mode (use chatComplete for simpler API)
  console.log('\nNon-Streaming Mode:');
  console.log('A: ');
  const nonStreamStart = Date.now();

  const response = await client.chat.chatComplete({
    message: question,
  });
  console.log(response.response);

  const nonStreamTime = Date.now() - nonStreamStart;
  console.log(`(Time: ${nonStreamTime}ms)`);

  console.log('\nStreaming provides better UX with instant feedback!');
}

async function main() {
  try {
    await streamingChatExample();
    await multiTurnConversation();
    await sessionManagement();
    await streamingVsNonStreaming();

    console.log('\n' + '='.repeat(70));
    console.log('  All streaming examples completed!');
    console.log('='.repeat(70));
  } catch (error) {
    console.error('\n❌ Error:', error);
    process.exit(1);
  }
}

main();
